Cldc 3114 enable support control of rent periods per organisation

arfacamble opened the pull request:
https://dluhcdigital.atlassian.net/browse/CLDC-3114

arfacamble committed:
write helper method to support having the correct rent period checkboxes checked

arfacamble committed:
update new and create in organisations controller and view to enable creation of relevant organisation rent periods
write tests for this

arfacamble committed:
small changes to models

arfacamble committed:
enable editing rent period in the UI
display change button on org details page
display rent periods question on edit page

arfacamble committed:
enable updating org rent periods
update logic in #update
related tests

arfacamble committed:
alter redirect after support user creates an organisation

arfacamble committed:
adjust various UI elements: ordering of rows, copy changes, label size and associated tests

arfacamble committed:
rework the #rent_period_labels method to return All under the correct conditions, rework tests related to that. + fix assorted tests that were either flakey or breaking due to addition of rent periods logic to create and update

arfacamble commented on app/views/organisations/show.html.erb:
the spec required the data agreement row to no longer be at the end

arfacamble committed:
amend failing tests and resolve linting complaints

arfacamble commented on spec/requests/OrganisationsController/organisations_controller_rent_periods_spec.rb:
```diff
@@ -0,0 +1,114 @@
+
+    it "the checkboxes for each rent period are checked by default" do
+      checkboxes = page.all "input[type='checkbox'][name='organisation[rent_periods][]']"
+      expect(checkboxes.count).to be > 5
```
interested in thoughts on this. It kind of looks random to be greater than 5 (which is an arbitrary number) but it's annoying when this type of test fails because the number of options changes by 1 due to something unrelated.

I think it's important to have something, otherwise the next line can pass when there are no checkboxes on the page.

maybe I should just change it to > 0 ? Or maybe it's good as is.

kosiakkatrina replied:
How likely would we add new rent periods?

arfacamble replied:
in what time frame?
low but not zero.
is this test currently doing its job and will it fail in future for good reason or no reason?

part of the reason why I've done this is the real expectation is the next line on 25, but any time you check all of something meet a condition the test can fail to be testing what it's meant to if there are none of that thing on the page because something earlier in the test setup gets broken at some point in the future so the test silently passes despite not doing what it's meant to

kosiakkatrina commented on app/controllers/organisations_controller.rb:
```diff
@@ -73,21 +73,30 @@ def details
-    @resource = Organisation.new(org_params)
-    if @resource.save
-      redirect_to organisations_path
+    selected_rent_periods = params.require(:organisation).permit(rent_periods: [])[:rent_periods].compact_blank
```
Would it make sense to extract this as a method?

arfacamble replied:
happy to.

kosiakkatrina commented on app/helpers/data_sharing_agreement_helper.rb:
```diff
@@ -1,7 +1,7 @@
   def data_sharing_agreement_row(user:, organisation:, summary_list:)
     summary_list.with_row do |row|
-      row.with_key { "Data Sharing Agreement" }
+      row.with_key { "Data protection agreement" }
```
Any reason for renaming it/is that the correct casing?

arfacamble replied:
It's just what is in the designs. I have messaged the higher ups to double check.

kosiakkatrina commented on app/helpers/organisations_helper.rb:
```diff
@@ -37,4 +38,14 @@ def organisation_name_row(user:, organisation:, summary_list:)
+      [code, period_copy]
+    end
+    rent_periods.to_h
+  end
```
Would this do the same thing?
```
  def rent_periods_with_checked_attr(checked_periods: nil)
    all_rent_periods = RentPeriod.rent_period_mappings
    rent_periods = all_rent_periods.each_with_object({}) do |(code, period), hash|
      hash[code] = period.merge(checked: checked_periods.nil? || checked_periods.include?(code))
    end
  end
  ```

arfacamble replied:
I was looking for an enumerator that returned a hash. thanks

arfacamble replied:
then I htought I would try with inject because that sounds even more like what we're doing and the linter told me to do what you did and explained why. all good learning.

kosiakkatrina commented on app/controllers/organisations_controller.rb:
```diff
@@ -94,6 +94,8 @@ def create
 
   def edit
     if current_user.data_coordinator? || current_user.support?
+      checked_periods = @organisation.organisation_rent_periods.map(&:rent_period).map(&:to_s)
```
I _think_ pluck(:rent_periods) might be more efficient

arfacamble replied:
very good yes. pretty sure you're right

kosiakkatrina commented on app/controllers/organisations_controller.rb:
```diff
@@ -133,7 +134,17 @@ def update
           flash[:notice] = I18n.t("organisation.updated")
         end
+        existing_rent_periods = @organisation.organisation_rent_periods.map(&:rent_period)
+        rent_periods_to_create = selected_rent_periods.map(&:to_i) - existing_rent_periods
```
Could we use `find_or_create_by` when creating the new periods instead?

arfacamble replied:
We probably could but I'm not really concerned about that.
I put a uniqueness validation in the model anyway, so I could just call create on them all and let the duplicate inserts fail quietly.
I'm more slightly annoyed to be getting things from the db before doing this anyway, which feels like it should be unnecessary. I could just pass all values through from the form in a hidden field and and delete all `OrganisationRentPeriod.where(org: @org, rent_period: [unchecked_vals])`
What are your thoughts on that?

kosiakkatrina reviewed (commented):
So I was under the impression that if no rent periods are selected then all the rent periods apply, is that not the case?
<img width="593" alt="image" src="https://github.com/communitiesuk/submit-social-housing-lettings-and-sales-data/assets/54268893/70c02cf6-36ba-4eaf-8ce0-7b6c0b47b5da">
I'd expect this to say "All rent periods" or something

Also, if a rent period was used but then it gets removed as part of the edit, what happens to the logs that used a that rent period?

arfacamble committed:
changes following review

arfacamble committed:
disable checkboxes for rent periods if they are in use so that users are not able to make existing logs invalid
hint text added to the question to explain this
I have also added all rent periods to a hidden field to remove the need to fetch them again form the db in the update method

arfacamble committed:
update validation to reflect the fact that an org having no associated rent periods no longer means they accept all rent periods
update tests adding both cases and removing unnecessary additional db additions

arfacamble committed:
rake task to create rent period associations for orgs that have none

arfacamble committed:
revert mistaken copy changes in designs

arfacamble committed:
create rent periods in factories as default, with an option to skip. skip automatic creation in tests specifically related to rent periods

arfacamble committed:
stub api call for factory value, update csv tests and fixtures accordingly

arfacamble committed:
extract a good chunk of tests out of lettings_log_spec and into a dedicated derived fields spec file. in many cases refactor tests

kosiakkatrina commented on spec/models/organisation_spec.rb:
```diff
@@ -242,6 +243,7 @@
 
   describe "scopes" do
     before do
+      described_class.destroy_all
```
How come we're now doing this here?

arfacamble replied:
I should be able to remove this now. This was related to what I have been talking about that I keep getting organisations floating around in the DB that aren't meant to be there

kosiakkatrina commented on spec/models/lettings_log_derived_fields_spec.rb:
```diff
@@ -0,0 +1,1178 @@
+RSpec.describe LettingsLog, type: :model do
+  before(:context) do
+    owning_organisation = build(:organisation)
+    @log = build(:lettings_log, :startdate_today, owning_organisation:, managing_organisation: owning_organisation)
```
Any reason we're not using let to define log? Looks like defining it this way in specs is an anti-pattern (?)

arfacamble replied:
this was required by doing before(:context) but I got rid of that now

kosiakkatrina commented on spec/models/lettings_log_derived_fields_spec.rb:
```diff
@@ -0,0 +1,1178 @@
+  describe "deriving has_benefits" do
+    it "correctly derives when the household receives benefits" do
+      benefits_codes = [1, 6, 8, 7]
+      @log.hb = benefits_codes.sample
```
nice

arfacamble commented on spec/models/lettings_log_derived_fields_spec.rb:
You will notice that the main weird thing that I am doing here is using `before(:context)` and therefore disabling a rubocop warning against doing this.
This means that one LettingsLog is instantiated and then reused throughout all tests in the file (there are some tests where this is not true below as clearing some derived values relies on calling some active record methods that check the value of a given field at the point the model was found in the DB).
using this before context then means the log has to be referred to with an instance variable to make it available in tests.

I'm doing this since I read somewhere that factories can even be quite time consuming to build (although the difference between create and build is much larger) but I intend to do some of my own testing now as to how longs tests take to run under conditions x, y, z

kosiakkatrina replied:
Would you be able to benchmark these tests and see how much time we're actually saving? Cause I think I'd prefer using let if it doesn't make much difference in time

arfacamble replied:
It was always my intention to benchmark, annoyingly with lots of weird behaviour in tests yesterday I didn't get to do that. I have now benchmarked this file against a refactor doing it the normal way and running them both ten times then it is faster, but only by 4% which I think is ABSOLUTELY NOT WORTH IT

arfacamble commented on spec/models/lettings_log_derived_fields_spec.rb:
```diff
@@ -0,0 +1,1178 @@
+
+      expect(@log.has_benefits).to be 0
+    end
+  end
```
how do we feel about stuff like this? There's a gentle smell to having any element of randomness in tests, but I feel like this finds the right balance between being concise and comprehensive

arfacamble commented on spec/models/lettings_log_spec.rb:
```diff
@@ -250,1841 +251,236 @@
+      expect(record_from_db["startdate"].day).to eq(10)
+      expect(record_from_db["startdate"].month).to eq(10)
+      expect(record_from_db["startdate"].year).to eq(2021)
+    end
```
sanity check please. These tests look silly to me, we save a date in the db, we call it back out, we can do standard ruby date things to it.
Can I delete?

arfacamble committed:
remove before(:context) and associated patterns.
use assign_attributes in various places for cleaner code

arfacamble committed:
escape . in regex for API call stubs to satisfy codeQL
remove destroy_all call at the start of a test that was dealing with leftover modesl in the test db

arfacamble committed:
further refactoring of various tests to reduce database interactions and improve speed

arfacamble committed:
remove outdated distinction between unitletas mappings from before 23/24

arfacamble committed:
remove tests that seem to be testing active record and/or ruby Date class
